types/types.go

```diff
@@ -56,13 +58,41 @@ func (h *Hash) Scan(src interface{}) error {
 	return nil
 }
 
-func (a Address) EIP55() string {
-	// TODO
-	return hex.EncodeToHex(a[:])
+// eip55 returns the checksummed address with 0x prefix
+// https://github.com/ethereum/EIPs/blob/master/EIPS/eip-55.md
+func (a Address) eip55() string {
```

Comment:
@Kourin1996 
There has got to be a way to make this parsing more efficient. Do you have any ideas?
I haven't benched it, but maybe we can use some golang tricks and slice optimizations.

Curious to hear your thoughts as well @lazartravica @Vuksan

Reply:
@zivkovicmilos 
If there are following premises, it's possible to optimize it
+ the length of given address and result address are same each other
+ the address consists only hex characters (0-9, A-F)

In this case, I think we can use a slice with fixed instead of dynamically concat. And we can compare character by comparison operator.

I wrote simple benchmarks and ran them. We may be able to optimize.

```go
package main

import (
	"strconv"
	"strings"
	"testing"
	"unicode"
)

const (
	OriginalAddr = "001d3f1ef827552ae1114027bd3ecf1f086ba0f9"
	HashedAddr   = "23a69c1653e4ebbb619b0b2cb8a9bad49892a8b9695d9a19d8f673ca991deae1"
	ExpectedAddr = "001d3F1ef827552Ae1114027BD3ECF1f086bA0F9"
)

// Based on original code
func ComvertBySimpleConcat() string {
	result := ""
	for idx, ch := range OriginalAddr {
		if ch >= '0' && ch <= '9' {
			result += string(ch)
		} else {
			hashedNibble, _ := strconv.ParseInt(string(HashedAddr[idx]), 16, 64)
			if hashedNibble > 7 {
				result += strings.ToUpper(string(ch))
			} else {
				result += string(ch)
			}
		}
	}
	return result
}

// Use fixed slice of rune
func ConvertWithFixedByteSlice() string {
	result := make([]rune, len(OriginalAddr))
	for idx, ch := range OriginalAddr {
		if ch >= '0' && ch <= '9' {
			result[idx] = ch
		} else {
			hashedNibble, _ := strconv.ParseInt(string(HashedAddr[idx]), 16, 64)
			if hashedNibble > 7 {
				result[idx] = unicode.ToUpper(ch)
			} else {
				result[idx] = ch
			}
		}
	}
	return string(result)
}

// Use fixed slice + optimized compare
func ComvertWithOptimizedCompare() string {
	result := make([]rune, len(OriginalAddr))
	for idx, ch := range OriginalAddr {
                // we expect ch and HashedAddr[idx] are any of hex character
		if ch >= '0' && ch <= '9' || HashedAddr[idx] >= '1' && HashedAddr[idx] <= '7' {
			result[idx] = ch
		} else {
			result[idx] = unicode.ToUpper(ch)
		}
	}
	return string(result)
}

func BenchmarkComvertBySimpleConcat(b *testing.B) {
	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		result := ComvertBySimpleConcat()
		if result != ExpectedAddr {
			panic("wrong result")
		}
	}
}

func BenchmarkConvertWithFixedByteSlice(b *testing.B) {
	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		result := ConvertWithFixedByteSlice()
		if result != ExpectedAddr {
			panic("wrong result")
		}
	}
}

func BenchmarkComvertWithOptimizedCompare(b *testing.B) {
	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		result := ComvertWithOptimizedCompare()
		if result != ExpectedAddr {
			panic("wrong result")
		}
	}
}
```

You can run this benchmark by this command:
```shell
$ go test -bench .
```

Result in my environment is as follows:
```shell
go test -bench .
goos: darwin
goarch: amd64
pkg: github.com/Kourin1996/benchmark1
BenchmarkComvertBySimpleConcat-8                  320055              3149 ns/op
BenchmarkConvertWithFixedByteSlice-8             1334090               912 ns/op
BenchmarkComvertWithOptimizedCompare-8           2982691               411 ns/op
PASS
ok      github.com/Kourin1996/benchmark1        5.126s
```

Reply:
Thank you for the detailed answer and for taking the time to benchmark different solutions 💯 

I was expecting a 1-2 sentence reply but you went above and beyond. Both of your assumptions are correct.
I've replaced the bulk of the eip55 method with the code you suggested with `ConvertWithOptimizedCompare` 👍

Reply:
Please rename this function to what it does.
Comment above it is enough for EIP origin IMO
```suggestion
func (a Address) checksumEncode() string {
```